lib/image_optim/cmd.rb

```diff
@@ -15,6 +17,40 @@ def run(*args)
         success
       end
 
+      def supports_timeout?
+        if defined?(JRUBY_VERSION)
+          JRUBY_VERSION >= '9.0.0.0'
+        else
+          RUBY_VERSION >= '1.9'
+        end
+      end
+
+      # Run the specified command, and kill it off if it runs longer
+      # than `timeout` seconds.
+      #
+      # Return success status
+      # Will raise an error when command timeouts
+      def run_with_timeout(timeout, *args)
+        return run(*args) unless timeout > 0 && supports_timeout?
+
+        success = false
+        init_options!(args)
+
+        begin
+          pid = Process.spawn(*args)
+          thread = Process.detach(pid)
+
+          if thread.join(timeout).nil?
+            cleanup_process(pid)
```

Comment:
@toy The reason I'm not calling `join` here is because we don't want the cleanup process to be blocking the main process.